mesa/model.py

```diff
@@ -71,39 +77,95 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         self.running = True
         self.schedule = None
         self.current_id = 0
-        self.agents_: defaultdict[type, dict] = defaultdict(dict)
+
+        self._setup_agent_registration()
 
         self._steps: int = 0
         self._time: TimeT = 0  # the model's clock
 
     @property
     def agents(self) -> AgentSet:
         """Provides an AgentSet of all agents in the model, combining agents from all types."""
-
-        if hasattr(self, "_agents"):
-            return self._agents
-        else:
-            all_agents = itertools.chain.from_iterable(self.agents_.values())
-            return AgentSet(all_agents, self)
+        return self._all_agents
 
     @agents.setter
     def agents(self, agents: Any) -> None:
         raise AttributeError(
-            "You are trying to set model.agents. In Mesa 3.0 and higher, this attribute will be "
+            "You are trying to set model.agents. In Mesa 3.0 and higher, this attribute is "
             "used by Mesa itself, so you cannot use it directly anymore."
             "Please adjust your code to use a different attribute name for custom agent storage."
         )
 
-        self._agents = agents
-
     @property
     def agent_types(self) -> list[type]:
         """Return a list of different agent types."""
-        return list(self.agents_.keys())
+        return list(self._agents_by_type.keys())
```

Comment:
The previous and the new change have different meaning altogether (list of all type of each agents vs list(set(agent types))). I suppose the docstring is ambiguous even though it is closer to the former meaning.

Reply:
That's not correct. In the old version `self.agents_` was the exact same datastructure as what is now `_agents_by_type`. In the old version: `self.agents_: defaultdict[type, dict] = defaultdict(dict)`. In the new version: `self._agents_by_type: dict[type, AgentSet] = {}`. So, in either case, keys is a list of types.

Reply:
That means it still stands that the docstring is not clear enough.

Reply:
This docstring has not been changed by me in this PR, so it is already in the existing code. If you suggest changing it, that would be fine, but why would that hold up this PR?

Also, what do you suggest changing? The return is a list, and this list contains the types of the agents in the model. The only, minor, ambiguity is that it is a list of the unique types of agents, not the type for each agent in the model.